### ISSUE TITLE
ICE while overriding default trait implementation

### ISSUE THREAD

**Issue by rohitjoshi** [rust-lang/rust#21716]
When I tried to override default implementation of `read_str()` method of json::Decoder trait,  I get error: internal compiler error: unexpected panic.  Ideally it should given me an error that you must implement all the trait methods.

    impl ::rustc_serialize::Decoder for AlertPayloadType {
        fn read_str(&mut self) -> Result<String, Decoder::Error> {
   		match self.pop() {
        	    Json::Null => Ok("".to_string()),
        	    value => Ok(format!("{}", value)),
    	       }
        }   
    }

**Versions:**
    rustc 1.0.0-nightly (458a6a2f6 2015-01-25 21:20:37 +0000)
    cargo 0.0.1-pre-nightly (bb28e71 2015-01-22 06:06:34 +0000)

**Back-trace:**
<pre>
apns rohitjoshi$ RUST_BACKTRACE=1 cargo build  --verbose
       Fresh log v0.2.0
       Fresh gcc v0.1.6
       Fresh libc v0.1.0
       Fresh rustc-serialize v0.2.9
       Fresh time v0.1.14
       Fresh net-utils v0.0.1 (https://github.com/rohitjoshi/net-utils.git#6ddc194e)
   Compiling apns v0.0.1 (file:///Users/rohitjoshi/projects/rust/apns)
     Running `rustc src/lib.rs --crate-name apns --crate-type lib -g -C metadata=1fcee4f6b418c0da -C extra-filename=-1fcee4f6b418c0da --out-dir /Users/rohitjoshi/projects/rust/apns/target --emit=dep-info,link -L dependency=/Users/rohitjoshi/projects/rust/apns/target -L dependency=/Users/rohitjoshi/projects/rust/apns/target/deps --extern rustc-serialize=/Users/rohitjoshi/projects/rust/apns/target/deps/librustc-serialize-8214bc2e838e96ca.rlib --extern net-utils=/Users/rohitjoshi/projects/rust/apns/target/deps/libnet-utils-6b70e46a05f1d859.rlib --extern log=/Users/rohitjoshi/projects/rust/apns/target/deps/liblog-4e79c2d7625e8c6f.rlib -L native=/Users/rohitjoshi/projects/rust/apns/target/build/time-24f9c2dbbbc6f19b/out`
ERROR:rbml::reader: failed to find block with tag 7
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: http://doc.rust-lang.org/complement-bugreport.html
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'explicit panic', /Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/src/librbml/lib.rs:262

stack backtrace:
   1:        0x105240d5c - sys::backtrace::write::h926cb97ba8fab334dbu
   2:        0x105264c8f - failure::on_fail::h85f0b0735db7566b8aB
   3:        0x1051cebae - rt::unwind::begin_unwind_inner::hf6d52be0daf8a2f44SA
   4:        0x1048260eb - rt::unwind::begin_unwind::h4549706688203109812
   5:        0x104824f89 - reader::get_doc::hb8b500b3799ad176nLa
   6:        0x10268d958 - metadata::decoder::item_type::h151361ae279f5ad3oVg
   7:        0x10269dddb - metadata::decoder::get_type::h399c3a212cddd10ax7g
   8:        0x10264ef5e - metadata::csearch::get_type::h304fcc3e5ae8f35anYj
   9:        0x101f8e4d4 - collect::CollectCtxt<'a, 'tcx>.AstConv<'tcx>::get_item_type_scheme::h0d60851b0297f9f5Sgu
  10:        0x101f7d691 - astconv::ast_path_to_ty::h112b74d43de603b0USs
  11:        0x101f814f0 - astconv::ast_ty_to_ty::unboxed_closure.30770
  12:        0x101f1c9f6 - astconv::ast_ty_to_ty::h6b3068569c155302Ejt
  13:        0x101f77629 - vec::Vec<T>.FromIterator<T>::from_iter::h9935643277956187684
  14:        0x101f757e6 - astconv::convert_angle_bracketed_parameters::h57532aaa51f9e9camts
  15:        0x101f1c106 - astconv::ast_path_substs_for_ty::h66e57c5063ab1bd7tis
  16:        0x101f7d78c - astconv::ast_path_to_ty::h112b74d43de603b0USs
  17:        0x101f814f0 - astconv::ast_ty_to_ty::unboxed_closure.30770
  18:        0x101f1c9f6 - astconv::ast_ty_to_ty::h6b3068569c155302Ejt
  19:        0x101f7817f - astconv::convert_ty_with_lifetime_elision::h11d3d7b7dc3faf4dtxs
  20:        0x101f830ff - astconv::ty_of_method_or_bare_fn::h2f9399575c080343uDt
  21:        0x101f82451 - astconv::ty_of_method::h2a9720f484e0bc45QBt
  22:        0x101fa5739 - collect::convert_methods::ty_of_method::h319fbc118542416bjFu
  23:        0x101f967e3 - collect::convert::h94c751b0a7ed02fdpIu
  24:        0x101f9a0d2 - visit::walk_item::h3534631790025090882
  25:        0x101fcc8b2 - check_crate::unboxed_closure.31907
  26:        0x101fca981 - check_crate::h911809ee92b3ea658My
  27:        0x1019b2ace - driver::phase_3_run_analysis_passes::hf23cd54e255d7d22VFa
  28:        0x101999ab6 - driver::compile_input::h3750206967d345b7Bba
  29:        0x101a587e6 - run_compiler::hc084e5c048185791R8b
  30:        0x101a55b33 - thunk::F.Invoke<A, R>::invoke::h3046682734678028285
  31:        0x101a54998 - rt::unwind::try::try_fn::h12824457894967850961
  32:        0x1052da3a9 - rust_try_inner
  33:        0x1052da396 - rust_try
  34:        0x101a54fd6 - thunk::F.Invoke<A, R>::invoke::h13746492197546964172
  35:        0x105250621 - sys::thread::thread_start::hb385b609f87fbce2r3w
  36:     0x7fff899342fc - _pthread_body
  37:     0x7fff89934279 - _pthread_body

Could not compile `apns`.

Caused by:
  Process didn't exit successfully: `rustc src/lib.rs --crate-name apns --crate-type lib -g -C metadata=1fcee4f6b418c0da -C extra-filename=-1fcee4f6b418c0da --out-dir /Users/rohitjoshi/projects/rust/apns/target --emit=dep-info,link -L dependency=/Users/rohitjoshi/projects/rust/apns/target -L dependency=/Users/rohitjoshi/projects/rust/apns/target/deps --extern rustc-serialize=/Users/rohitjoshi/projects/rust/apns/target/deps/librustc-serialize-8214bc2e838e96ca.rlib --extern net-utils=/Users/rohitjoshi/projects/rust/apns/target/deps/libnet-utils-6b70e46a05f1d859.rlib --extern log=/Users/rohitjoshi/projects/rust/apns/target/deps/liblog-4e79c2d7625e8c6f.rlib -L native=/Users/rohitjoshi/projects/rust/apns/target/build/time-24f9c2dbbbc6f19b/out` (status=101)

</pre>

**Comment by rohitjoshi** [comment]
I think found the issue. instead of defining `fn read_str(&mut self) -> DecodeResult<String> {}` , I defined as `fn read_str(&mut self) -> Result<String, Decoder::Error>{}`   

Here   `DecodeResult<String>` is same as  `Result<String, Decoder::Error>`  though.

**Comment by oli-obk** [comment]
smaller example:

```
struct Dummy;

impl Iterator for Dummy {
    type Item = i32;
    fn next(&mut self) -> Option<Iterator::Item> {
    }
}

fn main() {
}
```

**Comment by tamird** [comment]
No longer ICEs.
```shell
$ rustc main.rs
main.rs:5:34: 5:48 error: ambiguous associated type; specify the type using the syntax `<Type as core::iter::Iterator>::Item` [E0223]
main.rs:5     fn next(&mut self) -> Option<Iterator::Item> {
                                           ^~~~~~~~~~~~~~
error: aborting due to previous error
```
